packages/perspective-viewer-d3fc/src/ts/types.ts

```diff
@@ -0,0 +1,213 @@
+// ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
+// ┃ ██████ ██████ ██████       █      █      █      █      █ █▄  ▀███ █       ┃
+// ┃ ▄▄▄▄▄█ █▄▄▄▄▄ ▄▄▄▄▄█  ▀▀▀▀▀█▀▀▀▀▀ █ ▀▀▀▀▀█ ████████▌▐███ ███▄  ▀█ █ ▀▀▀▀▀ ┃
+// ┃ █▀▀▀▀▀ █▀▀▀▀▀ █▀██▀▀ ▄▄▄▄▄ █ ▄▄▄▄▄█ ▄▄▄▄▄█ ████████▌▐███ █████▄   █ ▄▄▄▄▄ ┃
+// ┃ █      ██████ █  ▀█▄       █ ██████      █      ███▌▐███ ███████▄ █       ┃
+// ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
+// ┃ Copyright (c) 2017, the Perspective Authors.                              ┃
+// ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
+// ┃ This file is part of the Perspective library, distributed under the terms ┃
+// ┃ of the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0). ┃
+// ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+
+import { IPerspectiveViewerPlugin } from "@finos/perspective-viewer";
+import { DataRow, Type } from "@finos/perspective";
+
+export interface Element {
+    msMatchesSelector(selectors: string): boolean;
+}
+
+export interface Chart {
```

Comment:
This could be very useful, but it's not used anywhere but ohlcCandle.
Also the proper types for the function are as follows:
```typescript
    (
    	// obtained from the results of `d3.select(this._container))` in `plugin._draw` 
        // resolves to d3.Selection<HTMLElement, unknown, null, undefined>,
        container: ReturnType<typeof d3.select<HTMLElement, unknown>>,
        settings: Settings
    ): void;
```
The container type here would cascade down and fill out all the container types throughout the codebase. They would probably need to replace the datum parameter with whatever datum they're using, but this seems to be the general form.

EDIT: I see why this isn't used - the current code just creates a function and plops a property on it. TS can't really handle that. I know the intention is to keep the code as close to the original as possible, but I think there's a pattern we could use to integrate this. This is similar to what ohlcCandle does, and I think we can do it for the other charts as well without much change. See [this playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwgFgQwE7AGIIMbAPZJFAXigAoAoKCqDbAO2AQEsaIkAuKAbwF8AacygM4RgwJgHMB7bnwCUhAHxQAbtgYATANylSTYCwBmmaPGTBO-CsDgNJsRCnRZcILZShgANgFcxTAPxSFm5QNAgAthDsAsBI4q7BFAD0iVDCGEFcWlzaye5IwhA0alBWNlBlCFBh2GpeHhCk9WYAHoQkMuwmKArmbk1Qhk54nfZomDh4bcTUdIzMSDxQQiLiAnIEitzxFP0Yo22DE-gIAlBeNADWNNgA7jRQJ3am21SjAHSePkxtHKER7AAiOAQEAAzJBfLALxIe57Z6kbKkGYCbD1N4ebBiYjNYgyRY4mQfby+GgyDRAA)
```typescript
type ChartFactory = (
    container: {},
    settings: {},
) => void;

interface Chart {
    this: ChartFactory;
    plugin?: {
        name: string;
        // etc
    };
}

// pretend this is a module
let x = (): Chart => {
    let factory: ChartFactory = (container, settings) => {};
    let chart = factory as unknown as Chart;
    chart.plugin = {name: "hey"};
    return chart;
}

console.log(x(), x().plugin);
```

Reply:
Good catch on the `container` arg type! I've updated the `Chart` type and chart functions. 

The `ChartFactory/Chart` pattern described above could be done, but I'm not 100% it needs to be (at least at the moment). There is some checking of the chart types where the constructors are being used:

```typescript
// src/ts/charts/charts.ts
const chartClasses: Chart[] = [
    barChart,
    columnChart,
    lineChart,
    xyLine,
    areaChart,
    yScatter,
    xyScatter,
    ohlc,
    candlestick,
    treemap,
    sunburst,
    heatmap,
];
```

So even though the type isn't defined on the functions themselves, they are "checked" on the chartClasses array type constraint. Is this enough for now? I dunno, curious what others think.

Reply:
> Is this enough for now? I dunno, curious what others think.

As I alluded to in the [first comment](https://github.com/finos/perspective/pull/2432#pullrequestreview-1743547897), this work needs to _eventually_ be all-or-nothing. Whether this is "enough" is the wrong question to ask - the real question is, what standard do you want to set for the developers who contribute after you? If someone else changes this code in ways that reduces type safety but the _contributor_ thinks its _enough_ and you don't, what do I merge? For a project of this scale, the only fair answer is that we encode our standards in tests, lint, CI, type-checking as much as we possibly can.

As I mentioned in my [subling comment](https://github.com/finos/perspective/pull/2432/files/f4c39d77d4567db54450f21142f3dd4b24d2bbfe#r1422019909) wrt to the constructor type of `SunburstSeries`, while this PR adds a lot to the _present_ developer experience, I am concerned that it is highly susceptible to bit-rot. Take this thread for example - `Chart` constraint only produces errors _on this line_, not in the definition of e.g. a mis-written `barChart` function; within that file, you can rename `plugin` (as it is marked nullable `?` in the type), add mystery fields, and do all sort of "broken" things without causing an error.

A good type system needs to prevent future developers from modifying the code incorrectly, not just provide intellisense for what is already written.